### ISSUE TITLE
Distribute without a go compiler

### ISSUE THREAD

**Issue by 4e554c4c** [4e554c4c/darkman.nvim#1]
One of the reasons to chose golang as a remote plugin is that it can produce static binaries that can be easily distributed without a runtime or dependencies.

Currently `darkman.nvim` requires a go compiler to be installed, so that `bin/darkman.nvim` can be compiled on-site.
Ideally, we should release "packaged" versions with a binary included, so that a go compiler is not necessary to install the plugin.

**Comment by 4e554c4c** [comment]
I created wbthomason/packer.nvim#1210 to discuss the possibility of downloading from a github release in packer

**Comment by 4e554c4c** [comment]
lazy.nvim is also unfriendly to this idea: folke/lazy.nvim#780

**Comment by 4e554c4c** [comment]
If something like packspec is used widely, we could switch to this: https://github.com/neovim/packspec

**Comment by eamonburns** [comment]
Have you considered distributing the plugin as a git repository, but also distributing the binary as a Github release? Then, the plugin could have a check somewhere around here:

https://github.com/4e554c4c/darkman.nvim/blob/2213b2b484606a20e260bb14b907586d7c6e7eaf/lua/darkman/init.lua#L23-L27

that makes sure the binary is present (and is the correct version). If it is not, it can download the binary from Github releases. If that fails (which should be rare), you can show an error that prompts the user to build the plugin.

**Comment by eamonburns** [comment]
One consideration would be that the check in Lua would have to stay in sync with the released binary. This is made a little easier due to the plugin being relatively stable.

**Comment by 4e554c4c** [comment]
> Have you considered distributing the plugin as a git repository, but also distributing the binary as a Github release?

Yes, that's the point of the issue, but the issue is package managers (like lazy) don't want to support this. Most package managers only `git clone` a repository. But github releases are archives like zip or tar. The linked issues are about lazy maybe being able to downoad a `zip`, but this was completely shut down

Some plugins get around that by downloading the release _themselves_ in a two-step process, but i'd rather not do this. i feel like it's out of scope for a tiny plugin to have to download its own binaries and do architecture detection and what-not.

**Comment by eamonburns** [comment]
> Some plugins get around that by downloading the release themselves in a two-step process

This is what I am asking (not the first half of your response).

> If [the binary] is not [present], it can download the binary from Github releases.

(quoting myself ^^^)

I would think it would be pretty simple to do with the new `vim.net.request`. In that case, you can say that a Go compiler and `build` "step" are required for Neovim 0.11 and older.

Pseudo-code:

```lua
...

local binPath = projectPath .. '/bin/darkman.nvim'

if not uv.fs_stat(binPath) then
  if vim.fn.has("nvim-0.12") == 1 then
    -- Look at required binary version, download/extract .tar.gz, move binary to `binPath`, clean up
    fetch_binary(binPath)
  else
    error("Binary is not present, and automatically fetching it is not supported for Neovim < 0.12. You must have a `build` step that compiles the binary (see README.md)")
  end
end

...
```

**Comment by 4e554c4c** [comment]
thanks for clarifying.
this still doesn't help for arch detection though. i use this plugin on two separate architectures (x64 and amd64) so it would need to download the correct release

**Comment by 4e554c4c** [comment]
it might be possible to use `jit.arch` for this, and have a different release for all supported: https://luajit.org/ext_jit.html

**Comment by eamonburns** [comment]
Yes, that is another thing to think about if you choose to download it manually (logic would be in `fetch_binary`).

It looks like you can also use `vim.uv.os_uname().machine` to get the architecture.

The good news is, you have the luxury of solving your (one) specific use case. The package managers have to solve for the general case. You can choose your artifact file name/URL format (e.g. `darkman.nvim_$os-$arch-%version.tar.gz`, where you know all the possible values of `$os`, `$arch`, and `$version`), and you can package a single binary. The package managers would have to have some way of formatting a URL based on all the different possible values for `$os`, `$arch`, and `$version` (e.g. on an x86_64 machine, `jit.arch` and `vim.uv.os_uname().machine` say different things: `x64` and `x86_64`).
